src/dotnet/commands/dotnet-build/IncrementalManager.cs

```diff
@@ -0,0 +1,201 @@
+
+            return versionsAreEqual
+                ? IncrementalResult.DoesNotNeedRebuild
+                : new IncrementalResult("the version or bitness of the CLI changed since the last build");
+        }
+
+        private IncrementalResult InputItemsChanged(ProjectGraphNode graphNode, CompilerIO compilerIO)
+        {
+            // check empty inputs / outputs
+            if (!compilerIO.Inputs.Any())
+            {
+                return new IncrementalResult("the project has no inputs");
+            }
+
+            if (!compilerIO.Outputs.Any())
+            {
+                return new IncrementalResult("the project has no outputs");
+            }
+
+            // check non existent items
+            var result = CheckMissingIO(compilerIO.Inputs, "inputs");
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            result = CheckMissingIO(compilerIO.Outputs, "outputs");
+            if (result.NeedsRebuilding)
+            {
+                return result;
+            }
+
+            // check cache against input glob pattern changes
+            var incrementalCacheFile = graphNode.ProjectContext.IncrementalCacheFile(_configuration, _buildBasePath, _outputPath);
+
+            if (!File.Exists(incrementalCacheFile))
+            {
+                // no cache present; cannot tell if anything changed
+                return IncrementalResult.DoesNotNeedRebuild;
+            }
+
+            var incrementalCache = IncrementalCache.ReadFromFile(incrementalCacheFile);
+
+            var diffResult = compilerIO.DiffInputs(incrementalCache.CompilerIO);
+
+            if (diffResult.Deletions.Any())
+            {
+                return new IncrementalResult("Input items removed from last build", diffResult.Deletions);
+            }
+
+            if (diffResult.Additions.Any())
+            {
+                return new IncrementalResult("Input items added from last build", diffResult.Additions);
+            }
+
+            return IncrementalResult.DoesNotNeedRebuild;
+        }
+
+        private IncrementalResult CheckMissingIO(IEnumerable<string> items, string itemsType)
+        {
+            var missingItems = items.Where(i => !File.Exists(i)).ToList();
+
+            return missingItems.Any()
+                ? new IncrementalResult($"expected {itemsType} are missing", missingItems)
+                : IncrementalResult.DoesNotNeedRebuild;
+        }
+
+        private IncrementalResult TimestampsChanged(CompilerIO compilerIO)
+        {
+            // find the output with the earliest write time
+            var minDateUtc = File.GetLastWriteTimeUtc(compilerIO.Outputs.First());
+
+            foreach (var outputPath in compilerIO.Outputs)
+            {
+                if (File.GetLastWriteTimeUtc(outputPath) >= minDateUtc)
```

Comment:
Going to the file system (or any other I/O) is never an "insignificant" performing action.